# Análisis de la Calidad del Aire en el Valle de México

## Módulo: Procesamiento de Datos

### Lectura y Análisis exploratorio de los datos

In [181]:
#Importando paquetes
import pandas as pd
import numpy as np

In [182]:
#Leyendo un archivo de datos IMECA para averiguar que contiene
df=pd.read_csv("datos_IMECA/csv/imeca2005.csv")
df

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2005     1              19                          19   
1     01/01/2005     2              19                          19   
2     01/01/2005     3              17                          19   
3     01/01/2005     4              16                          19   
4     01/01/2005     5              16                          19   
...          ...   ...             ...                         ...   
8755  31/12/2005    20              15                          10   
8756  31/12/2005    21              12                          10   
8757  31/12/2005    22               9                          10   
8758  31/12/2005    23               8                          11   
8759  31/12/2005    24               9                          11   

      Noroeste dioxido de nitrogeno  Noroeste monoxido de carbono  \
0                                43                            42   
1                                42                            45   
2                                42                            48   
3                                42                            52   
4                                39                            53   
...                             ...                           ...   
8755                             38                            18   
8756                             34                            20   
8757                             31                            23   
8758                             25                            27   
8759                             23                            30   

      Noroeste PM10  Noreste Ozono  Noreste dioxido de azufre  \
0                80             10                         22   
1                84              7                         22   
2                88              5                         22   
3                91              6                         22   
4                94              8                         23   
...             ...            ...                        ...   
8755             61             41                         12   
8756             61             35                         12   
8757             61             28                         12   
8758             62             23                         12   
8759             64             22                         12   

      Noreste dioxido de nitrogeno  ...  Suroeste Ozono  \
0                               34  ...              26   
1                               35  ...              19   
2                               34  ...              23   
3                               33  ...              21   
4                               34  ...              23   
...                            ...  ...             ...   
8755                            34  ...              33   
8756                            29  ...              27   
8757                            33  ...              19   
8758                            31  ...              22   
8759                            31  ...              20   

      Suroeste dioxido de azufre  Suroeste dioxido de nitrogeno  \
0                             12                             39   
1                             12                             34   
2                             12                             33   
3                             12                             32   
4                             12                             31   
...                          ...                            ...   
8755                           5                             24   
8756                           5                             30   
8757                           5                             30   
8758                           5                             25   
8759                           5                             24   

      Suroeste monoxido de carbono  Suroeste P

In [3]:
#Revisando dimensión
df.shape

(8760, 27)

In [183]:
#Revisando columnas que contiene
df.columns

Index(['Fecha', 'Hora', 'Noroeste Ozono', 'Noroeste dioxido de azufre',
       'Noroeste dioxido de nitrogeno', 'Noroeste monoxido de carbono',
       'Noroeste PM10', 'Noreste Ozono', 'Noreste dioxido de azufre',
       'Noreste dioxido de nitrogeno', 'Noreste monoxido de carbono',
       'Noreste PM10', 'Centro Ozono', 'Centro dioxido de azufre',
       'Centro dioxido de nitrogeno', 'Centro monoxido de carbono',
       'Centro PM10', 'Suroeste Ozono', 'Suroeste dioxido de azufre',
       'Suroeste dioxido de nitrogeno', 'Suroeste monoxido de carbono',
       'Suroeste PM10', 'Sureste Ozono', 'Sureste dioxido de azufre',
       'Sureste dioxido de nitrogeno', 'Sureste monoxido de carbono',
       'Sureste PM10'],
      dtype='object')

Tenemos un archivo de datos por año, vamos a unir todos en un solo DataFrame para trabajar con ellos más facilmente

In [184]:
#primero creamos una lista que contenga todas las rutas de acceso a los archivos csv
lista=[ f'datos_IMECA/csv/imeca{2005+i}.csv' for i in range(0,16)]
lista

['datos_IMECA/csv/imeca2005.csv',
 'datos_IMECA/csv/imeca2006.csv',
 'datos_IMECA/csv/imeca2007.csv',
 'datos_IMECA/csv/imeca2008.csv',
 'datos_IMECA/csv/imeca2009.csv',
 'datos_IMECA/csv/imeca2010.csv',
 'datos_IMECA/csv/imeca2011.csv',
 'datos_IMECA/csv/imeca2012.csv',
 'datos_IMECA/csv/imeca2013.csv',
 'datos_IMECA/csv/imeca2014.csv',
 'datos_IMECA/csv/imeca2015.csv',
 'datos_IMECA/csv/imeca2016.csv',
 'datos_IMECA/csv/imeca2017.csv',
 'datos_IMECA/csv/imeca2018.csv',
 'datos_IMECA/csv/imeca2019.csv',
 'datos_IMECA/csv/imeca2020.csv']

In [185]:
#Realizamos la lectura de todos los archivos csv juntos y reemplazamos -99 por NaN
DF_list = [(pd.read_csv(f'{i}',sep=',')).replace(-99, np.nan) for i in lista]

In [186]:
#comprobamos que se puedan leer de forma correcta los datos
DF_list[0]

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2005     1              19                          19   
1     01/01/2005     2              19                          19   
2     01/01/2005     3              17                          19   
3     01/01/2005     4              16                          19   
4     01/01/2005     5              16                          19   
...          ...   ...             ...                         ...   
8755  31/12/2005    20              15                          10   
8756  31/12/2005    21              12                          10   
8757  31/12/2005    22               9                          10   
8758  31/12/2005    23               8                          11   
8759  31/12/2005    24               9                          11   

      Noroeste dioxido de nitrogeno  Noroeste monoxido de carbono  \
0                                43                            42   
1                                42                            45   
2                                42                            48   
3                                42                            52   
4                                39                            53   
...                             ...                           ...   
8755                             38                            18   
8756                             34                            20   
8757                             31                            23   
8758                             25                            27   
8759                             23                            30   

      Noroeste PM10  Noreste Ozono  Noreste dioxido de azufre  \
0                80             10                         22   
1                84              7                         22   
2                88              5                         22   
3                91              6                         22   
4                94              8                         23   
...             ...            ...                        ...   
8755             61             41                         12   
8756             61             35                         12   
8757             61             28                         12   
8758             62             23                         12   
8759             64             22                         12   

      Noreste dioxido de nitrogeno  ...  Suroeste Ozono  \
0                               34  ...              26   
1                               35  ...              19   
2                               34  ...              23   
3                               33  ...              21   
4                               34  ...              23   
...                            ...  ...             ...   
8755                            34  ...              33   
8756                            29  ...              27   
8757                            33  ...              19   
8758                            31  ...              22   
8759                            31  ...              20   

      Suroeste dioxido de azufre  Suroeste dioxido de nitrogeno  \
0                             12                             39   
1                             12                             34   
2                             12                             33   
3                             12                             32   
4                             12                             31   
...                          ...                            ...   
8755                           5                             24   
8756                           5                             30   
8757                           5                             30   
8758                           5                             25   
8759                           5                             24   

      Suroeste monoxido de carbono  Suroeste P

In [187]:
#segunda comprobación
DF_list[13]

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2018     1            24.0                           8   
1     01/01/2018     2            23.0                           8   
2     01/01/2018     3            18.0                           8   
3     01/01/2018     4            12.0                          10   
4     01/01/2018     5            11.0                          12   
...          ...   ...             ...                         ...   
8755  31/12/2018    20            20.0                           5   
8756  31/12/2018    21            12.0                           5   
8757  31/12/2018    22             6.0                           5   
8758  31/12/2018    23             4.0                           5   
8759  31/12/2018    24             NaN                           5   

      Noroeste dioxido de nitrogeno  Noroeste monoxido de carbono  \
0                              14.0                          12.0   
1                              18.0                          12.0   
2                              19.0                          12.0   
3                              18.0                          12.0   
4                              18.0                          12.0   
...                             ...                           ...   
8755                           26.0                           5.0   
8756                           26.0                           6.0   
8757                           26.0                           7.0   
8758                           22.0                           8.0   
8759                            NaN                          12.0   

      Noroeste PM10  Noreste Ozono  Noreste dioxido de azufre  \
0               106           24.0                          5   
1               107           29.0                          5   
2               108           23.0                          5   
3               110           18.0                          5   
4               113           17.0                          5   
...             ...            ...                        ...   
8755             46           17.0                         12   
8756             49           14.0                         12   
8757             50           11.0                         12   
8758             54            6.0                         12   
8759             61            NaN                         12   

      Noreste dioxido de nitrogeno  ...  Suroeste dioxido de nitrogeno  \
0                             18.0  ...                           17.0   
1                             16.0  ...                           12.0   
2                             19.0  ...                           10.0   
3                             17.0  ...                            9.0   
4                             18.0  ...                            9.0   
...                            ...  ...                            ...   
8755                          29.0  ...                           18.0   
8756                          20.0  ...                           25.0   
8757                          21.0  ...                           26.0   
8758                          21.0  ...                           21.0   
8759                           NaN  ...                            NaN   

      Suroeste monoxido de carbono  Suroeste PM10  Sureste Ozono  \
0                              8.0             67            5.0   
1                              7.0             70            8.0   
2                              7.0             73            4.0   
3                              7.0             76            6.0   
4                              6.0             77            4.0   
...                            ...            ...            ...   
8755                           4.0             26           16.0   
8756                           5.0             28           11.0   
8757                           6.0             30          

In [188]:
#El Dataframe número 14 y 15 tienen 32 columnas cuando lo correcto serían 27
#Eliminamos las columnas que sobran
cols = [7,13,19,25,31]
DF_list[14]=DF_list[14].drop(DF_list[14].columns[cols],axis=1)
DF_list[15]=DF_list[15].drop(DF_list[15].columns[cols],axis=1)

In [190]:
#Se puede observar que existe un inconveniente: por alguna razón algunos DataFrames 
#tienen columnas de más, que contienen solo elementos vacíos, es necesario eliminarlas
DF_list[12].columns

Index(['Fecha', 'Hora', 'Noroeste Ozono', 'Noroeste dioxido de azufre',
       'Noroeste dioxido de nitrogeno', 'Noroeste monoxido de carbono',
       'Noroeste PM10', 'Noreste Ozono', 'Noreste dioxido de azufre',
       'Noreste dioxido de nitrogeno', 'Noreste monoxido de carbono',
       'Noreste PM10', 'Centro Ozono', 'Centro dioxido de azufre',
       'Centro dioxido de nitrogeno', 'Centro monoxido de carbono',
       'Centro PM10', 'Suroeste Ozono', 'Suroeste dioxido de azufre',
       'Suroeste dioxido de nitrogeno', 'Suroeste monoxido de carbono',
       'Suroeste PM10', 'Sureste Ozono', 'Sureste dioxido de azufre',
       'Sureste dioxido de nitrogeno', 'Sureste monoxido de carbono',
       'Sureste PM10', 'Unnamed: 27', 'Unnamed: 28'],
      dtype='object')

In [191]:
DF_list[12].shape

(8760, 29)

In [192]:
DF_list[12]['Unnamed: 27'] #Se observa que contiene puros elementos vacíos

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
        ..
8755   NaN
8756   NaN
8757   NaN
8758   NaN
8759   NaN
Name: Unnamed: 27, Length: 8760, dtype: float64

In [193]:
#ahora eliminamos las columnas que aparecen como sobrantes
DF_list2=[i.dropna(axis="columns", how='all') for i in DF_list]
DF_list2[0]

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2005     1              19                          19   
1     01/01/2005     2              19                          19   
2     01/01/2005     3              17                          19   
3     01/01/2005     4              16                          19   
4     01/01/2005     5              16                          19   
...          ...   ...             ...                         ...   
8755  31/12/2005    20              15                          10   
8756  31/12/2005    21              12                          10   
8757  31/12/2005    22               9                          10   
8758  31/12/2005    23               8                          11   
8759  31/12/2005    24               9                          11   

      Noroeste dioxido de nitrogeno  Noroeste monoxido de carbono  \
0                                43                            42   
1                                42                            45   
2                                42                            48   
3                                42                            52   
4                                39                            53   
...                             ...                           ...   
8755                             38                            18   
8756                             34                            20   
8757                             31                            23   
8758                             25                            27   
8759                             23                            30   

      Noroeste PM10  Noreste Ozono  Noreste dioxido de azufre  \
0                80             10                         22   
1                84              7                         22   
2                88              5                         22   
3                91              6                         22   
4                94              8                         23   
...             ...            ...                        ...   
8755             61             41                         12   
8756             61             35                         12   
8757             61             28                         12   
8758             62             23                         12   
8759             64             22                         12   

      Noreste dioxido de nitrogeno  ...  Suroeste Ozono  \
0                               34  ...              26   
1                               35  ...              19   
2                               34  ...              23   
3                               33  ...              21   
4                               34  ...              23   
...                            ...  ...             ...   
8755                            34  ...              33   
8756                            29  ...              27   
8757                            33  ...              19   
8758                            31  ...              22   
8759                            31  ...              20   

      Suroeste dioxido de azufre  Suroeste dioxido de nitrogeno  \
0                             12                             39   
1                             12                             34   
2                             12                             33   
3                             12                             32   
4                             12                             31   
...                          ...                            ...   
8755                           5                             24   
8756                           5                             30   
8757                           5                             30   
8758                           5                             25   
8759                           5                             24   

      Suroeste monoxido de carbono  Suroeste P

In [194]:
#observamos que las columnas sobrantes se fueron
DF_list2[12].columns

Index(['Fecha', 'Hora', 'Noroeste Ozono', 'Noroeste dioxido de azufre',
       'Noroeste dioxido de nitrogeno', 'Noroeste monoxido de carbono',
       'Noroeste PM10', 'Noreste Ozono', 'Noreste dioxido de azufre',
       'Noreste dioxido de nitrogeno', 'Noreste monoxido de carbono',
       'Noreste PM10', 'Centro Ozono', 'Centro dioxido de azufre',
       'Centro dioxido de nitrogeno', 'Centro monoxido de carbono',
       'Centro PM10', 'Suroeste Ozono', 'Suroeste dioxido de azufre',
       'Suroeste dioxido de nitrogeno', 'Suroeste monoxido de carbono',
       'Suroeste PM10', 'Sureste Ozono', 'Sureste dioxido de azufre',
       'Sureste dioxido de nitrogeno', 'Sureste monoxido de carbono',
       'Sureste PM10'],
      dtype='object')

In [195]:
#Revisamos que todos los DataFrame tengan el mismo tamaño
for i in range(0,16):
    print(np.shape(DF_list2[i]))

(8760, 27)
(8760, 27)
(8760, 27)
(8784, 27)
(8760, 27)
(8760, 27)
(8760, 27)
(8784, 27)
(8781, 27)
(8760, 27)
(8760, 27)
(8784, 27)
(8760, 27)
(8760, 27)
(8760, 27)
(8784, 27)


Se observa que hay años en los que se tiene 8784 filas en lugar de 8760, esto es porque se trata de años bisiestos y tienen un día más, lo que equivale a 24 observaciones extra.

In [196]:
#Renombramos todas las columnas para no generar problemas al momento de concatenar
#DF_list2[0].columns

nombres_columnas=( "fecha"           ,        "hora"                ,          "noroeste_ozono"               ,
 "noroeste_dioxido_de_azufre"   , "noroeste_dioxido_de_nitrogeno", "noroeste_monoxido_de_carbono" ,
 "noroeste_pm10"                , "noreste_ozono"                , "noreste_dioxido_de_azufre"    ,
 "noreste_dioxido_de_nitrogeno" , "noreste_monoxido_de_carbono"  , "noreste_pm10"                 ,
 "centro_ozono"                 , "centro_dioxido_de_azufre"     , "centro_dioxido_de_nitrogeno"  ,
 "centro_monoxido_de_carbono"   , "centro_pm10"                  , "suroeste_ozono"               ,
 "suroeste_dioxido_de_azufre"   , "suroeste_dioxido_de_nitrogeno", "suroeste_monoxido_de_carbono" ,
 "suroeste_pm10"                , "sureste_ozono"                , "sureste_dioxido_de_azufre"    ,
 "sureste_dioxido_de_nitrogeno" , "sureste_monoxido_de_carbono"  , "sureste_pm10" )

for i in DF_list2:
    i.columns = nombres_columnas

In [197]:
#Concatenando todo en un mismo DataFrame
DF_all = pd.concat(DF_list2,axis=0)
DF_all=DF_all.reset_index(drop=True)
DF_all

fecha  hora  noroeste_ozono  noroeste_dioxido_de_azufre  \
0       01/01/2005   1.0            19.0                        19.0   
1       01/01/2005   2.0            19.0                        19.0   
2       01/01/2005   3.0            17.0                        19.0   
3       01/01/2005   4.0            16.0                        19.0   
4       01/01/2005   5.0            16.0                        19.0   
...            ...   ...             ...                         ...   
140272  31/12/2020  20.0            25.0                         4.0   
140273  31/12/2020  21.0            22.0                         4.0   
140274  31/12/2020  22.0            18.0                         4.0   
140275  31/12/2020  23.0            16.0                         4.0   
140276  31/12/2020  24.0            16.0                         4.0   

        noroeste_dioxido_de_nitrogeno  noroeste_monoxido_de_carbono  \
0                                43.0                          42.0   
1                                42.0                          45.0   
2                                42.0                          48.0   
3                                42.0                          52.0   
4                                39.0                          53.0   
...                               ...                           ...   
140272                            9.0                           2.0   
140273                            9.0                           2.0   
140274                           18.0                           3.0   
140275                           18.0                           5.0   
140276                           18.0                           5.0   

        noroeste_pm10  noreste_ozono  noreste_dioxido_de_azufre  \
0                80.0           10.0                       22.0   
1                84.0            7.0                       22.0   
2                88.0            5.0                       22.0   
3                91.0            6.0                       22.0   
4                94.0            8.0                       23.0   
...               ...            ...                        ...   
140272           88.0           14.0                        4.0   
140273           88.0           11.0                        4.0   
140274           88.0           10.0                        4.0   
140275           88.0            9.0                        4.0   
140276           88.0           11.0                        4.0   

        noreste_dioxido_de_nitrogeno  ...  suroeste_ozono  \
0                               34.0  ...            26.0   
1                               35.0  ...            19.0   
2                               34.0  ...            23.0   
3                               33.0  ...            21.0   
4                               34.0  ...            23.0   
...                              ...  ...             ...   
140272                          15.0  ...            25.0   
140273                          17.0  ...            16.0   
140274                          17.0  ...             9.0   
140275                          19.0  ...            11.0   
140276                          18.0  ...            16.0   

        suroeste_dioxido_de_azufre  suroeste_dioxido_de_nitrogeno  \
0                             12.0                           39.0   
1                             12.0                           34.0   
2                             12.0                           33.0   
3                             12.0                           32.0   
4                             12.0                           31.0   
...                            ...                            ...   
140272                         2.0                           14.0   
140273                         2.0                           15.0   
140274                         2.0                           16.0   
140275                         2.0                           15.0  

### Limpieza NaNs

In [198]:
#Revisando datos faltantes por columna
DF_all.isna().sum()

fecha                              21
hora                               21
noroeste_ozono                   1306
noroeste_dioxido_de_azufre         22
noroeste_dioxido_de_nitrogeno    1297
noroeste_monoxido_de_carbono     2027
noroeste_pm10                     174
noreste_ozono                    1299
noreste_dioxido_de_azufre          21
noreste_dioxido_de_nitrogeno     1331
noreste_monoxido_de_carbono      2073
noreste_pm10                      369
centro_ozono                     1319
centro_dioxido_de_azufre          101
centro_dioxido_de_nitrogeno      1366
centro_monoxido_de_carbono       2070
centro_pm10                       555
suroeste_ozono                   1236
suroeste_dioxido_de_azufre         42
suroeste_dioxido_de_nitrogeno    1261
suroeste_monoxido_de_carbono     2050
suroeste_pm10                     365
sureste_ozono                    1319
sureste_dioxido_de_azufre         126
sureste_dioxido_de_nitrogeno     1485
sureste_monoxido_de_carbono      2231
sureste_pm10

Se nota que hay columnas que presentan más datos faltantes que otras. Puesto que no podemos despreciar las mediciones de toda la fila solo porque una observación en una columna es nula, optamos en su lugar por llenar los valores vacíos con el valor promedio de la columna. De esta forma nos quedamos con suficientes datos para trabajar y mantenemos su valor promedio.

In [199]:
#Llenando los valores numéricos con el valor promedio
DF_all_subset=DF_all.loc[:,"noroeste_ozono":"sureste_pm10"]

for i in DF_all_subset.columns:
    DF_all[i]=DF_all[i].fillna(DF_all[i].mean())

In [126]:
#Corroborando que se hayan reemplazado todos los datos faltantes numéricos
DF_all.isna().sum()

fecha                            21
hora                             21
noroeste_ozono                    0
noroeste_dioxido_de_azufre        0
noroeste_dioxido_de_nitrogeno     0
noroeste_monoxido_de_carbono      0
noroeste_pm10                     0
noreste_ozono                     0
noreste_dioxido_de_azufre         0
noreste_dioxido_de_nitrogeno      0
noreste_monoxido_de_carbono       0
noreste_pm10                      0
centro_ozono                      0
centro_dioxido_de_azufre          0
centro_dioxido_de_nitrogeno       0
centro_monoxido_de_carbono        0
centro_pm10                       0
suroeste_ozono                    0
suroeste_dioxido_de_azufre        0
suroeste_dioxido_de_nitrogeno     0
suroeste_monoxido_de_carbono      0
suroeste_pm10                     0
sureste_ozono                     0
sureste_dioxido_de_azufre         0
sureste_dioxido_de_nitrogeno      0
sureste_monoxido_de_carbono       0
sureste_pm10                      0
dtype: int64

In [201]:
#Revisando valores nulos de fechas
DF_all[DF_all["fecha"].isna()]

fecha  hora  noroeste_ozono  noroeste_dioxido_de_azufre  \
78888   NaN   NaN       28.903253                    9.383715   
78889   NaN   NaN       28.903253                    9.383715   
78890   NaN   NaN       28.903253                    9.383715   
78891   NaN   NaN       28.903253                    9.383715   
78892   NaN   NaN       28.903253                    9.383715   
78893   NaN   NaN       28.903253                    9.383715   
78894   NaN   NaN       28.903253                    9.383715   
78895   NaN   NaN       28.903253                    9.383715   
78896   NaN   NaN       28.903253                    9.383715   
78897   NaN   NaN       28.903253                    9.383715   
78898   NaN   NaN       28.903253                    9.383715   
78899   NaN   NaN       28.903253                    9.383715   
78900   NaN   NaN       28.903253                    9.383715   
78901   NaN   NaN       28.903253                    9.383715   
78902   NaN   NaN       28.903253                    9.383715   
78903   NaN   NaN       28.903253                    9.383715   
78904   NaN   NaN       28.903253                    9.383715   
78905   NaN   NaN       28.903253                    9.383715   
78906   NaN   NaN       28.903253                    9.383715   
78907   NaN   NaN       28.903253                    9.383715   
78908   NaN   NaN       28.903253                    9.383715   

       noroeste_dioxido_de_nitrogeno  noroeste_monoxido_de_carbono  \
78888                      17.478565                     10.089902   
78889                      17.478565                     10.089902   
78890                      17.478565                     10.089902   
78891                      17.478565                     10.089902   
78892                      17.478565                     10.089902   
78893                      17.478565                     10.089902   
78894                      17.478565                     10.089902   
78895                      17.478565                     10.089902   
78896                      17.478565                     10.089902   
78897                      17.478565                     10.089902   
78898                      17.478565                     10.089902   
78899                      17.478565                     10.089902   
78900                      17.478565                     10.089902   
78901                      17.478565                     10.089902   
78902                      17.478565                     10.089902   
78903                      17.478565                     10.089902   
78904                      17.478565                     10.089902   
78905                      17.478565                     10.089902   
78906                      17.478565                     10.089902   
78907                      17.478565                     10.089902   
78908                      17.478565                     10.089902   

       noroeste_pm10  noreste_ozono  noreste_dioxido_de_azufre  \
78888      55.736244      31.219862                   7.527464   
78889      55.736244      31.219862                   7.527464   
78890      55.736244      31.219862                   7.527464   
78891      55.736244      31.219862                   7.527464   
78892      55.736244      31.219862                   7.527464   
78893      55.736244      31.219862                   7.527464   
78894      55.736244      31.219862                   7.527464   
78895      55.736244      31.219862                   7.527464   
78896      55.736244      31.219862                   7.527464   
78897      55.736244      31.219862                   7.527464   
78898      55.736244      31.219862                   7.527464   
78899      55.736244      31.219862                   7.527464   
78900      55.736244      31.219862                   7.527464   
78901      55.736244      31.219862                   7.527464   
78902      55.736244      31.219862         

In [202]:
#Revisando la última fecha correcta
DF_all.loc[78887]

fecha                            31/12/2013
hora                                     24
noroeste_ozono                            9
noroeste_dioxido_de_azufre               22
noroeste_dioxido_de_nitrogeno            15
noroeste_monoxido_de_carbono              7
noroeste_pm10                            48
noreste_ozono                             7
noreste_dioxido_de_azufre                11
noreste_dioxido_de_nitrogeno             15
noreste_monoxido_de_carbono              12
noreste_pm10                             48
centro_ozono                              1
centro_dioxido_de_azufre                 16
centro_dioxido_de_nitrogeno              16
centro_monoxido_de_carbono                9
centro_pm10                              48
suroeste_ozono                            5
suroeste_dioxido_de_azufre               15
suroeste_dioxido_de_nitrogeno            20
suroeste_monoxido_de_carbono             13
suroeste_pm10                            51
sureste_ozono                   

In [203]:
#Revisando fecha posterior a la última vacía
DF_all.loc[78909]

fecha                            01/01/2014
hora                                      1
noroeste_ozono                            8
noroeste_dioxido_de_azufre               18
noroeste_dioxido_de_nitrogeno            15
noroeste_monoxido_de_carbono              7
noroeste_pm10                            51
noreste_ozono                             8
noreste_dioxido_de_azufre                11
noreste_dioxido_de_nitrogeno             15
noreste_monoxido_de_carbono              13
noreste_pm10                             48
centro_ozono                              1
centro_dioxido_de_azufre                 15
centro_dioxido_de_nitrogeno              15
centro_monoxido_de_carbono                9
centro_pm10                              48
suroeste_ozono                            4
suroeste_dioxido_de_azufre               14
suroeste_dioxido_de_nitrogeno            18
suroeste_monoxido_de_carbono             14
suroeste_pm10                            53
sureste_ozono                   

In [204]:
#Se corrobora que el archivo de 2014 esté correcto
pd.read_csv("datos_IMECA/csv/imeca2014.csv")

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2014     1               8                          18   
1     01/01/2014     2               8                          18   
2     01/01/2014     3               5                          18   
3     01/01/2014     4               2                          18   
4     01/01/2014     5               3                          18   
...          ...   ...             ...                         ...   
8755  31/12/2014    20              27                          25   
8756  31/12/2014    21              18                          25   
8757  31/12/2014    22              12                          25   
8758  31/12/2014    23               9                          26   
8759  31/12/2014    24              12                          26   

      Noroeste dioxido de nitrogeno  Noroeste monoxido de carbono  \
0                                15                             7   
1                                14                             7   
2                                14                             7   
3                                14                             9   
4                                13                            11   
...                             ...                           ...   
8755                             31                            13   
8756                             29                            12   
8757                             30                            12   
8758                             27                            13   
8759                             25                            14   

      Noroeste PM10  Noreste Ozono  Noreste dioxido de azufre  \
0                51              8                         11   
1                52              8                         12   
2                53              8                         13   
3                53              7                         14   
4                56              5                         15   
...             ...            ...                        ...   
8755            111             26                         34   
8756            111             13                         34   
8757            111              6                         34   
8758            111              4                         35   
8759            111              3                         34   

      Noreste dioxido de nitrogeno  ...  Suroeste dioxido de nitrogeno  \
0                               15  ...                             18   
1                               14  ...                             16   
2                               13  ...                             15   
3                               13  ...                             14   
4                               12  ...                             13   
...                            ...  ...                            ...   
8755                            44  ...                             61   
8756                            32  ...                             49   
8757                            29  ...                             49   
8758                            23  ...                             45   
8759                            19  ...                             34   

      Suroeste monoxido de carbono  Suroeste PM10  Sureste Ozono  \
0                               14             53              3   
1                               14             53              5   
2                               13             53              4   
3                               13             54              4   
4                               14             56              4   
...                            ...            ...            ...   
8755                            19            116             76   
8756                            20            116             24   
8757                            22            117          

In [205]:
#Observando el archivo del 2013
pd.read_csv("datos_IMECA/csv/imeca2013.csv")

Fecha  Hora  Noroeste Ozono  Noroeste dioxido de azufre  \
0     01/01/2013   1.0             3.0                         2.0   
1     01/01/2013   2.0             3.0                         3.0   
2     01/01/2013   3.0             4.0                         3.0   
3     01/01/2013   4.0             5.0                         3.0   
4     01/01/2013   5.0             9.0                         4.0   
...          ...   ...             ...                         ...   
8776         NaN   NaN             NaN                         NaN   
8777         NaN   NaN             NaN                         NaN   
8778         NaN   NaN             NaN                         NaN   
8779         NaN   NaN             NaN                         NaN   
8780         NaN   NaN             NaN                         NaN   

      Noroeste dioxido de nitrogeno  Noroeste monoxido de carbono  \
0                              17.0                          10.0   
1                              16.0                          12.0   
2                              16.0                          15.0   
3                              15.0                          17.0   
4                              16.0                          20.0   
...                             ...                           ...   
8776                            NaN                           NaN   
8777                            NaN                           NaN   
8778                            NaN                           NaN   
8779                            NaN                           NaN   
8780                            NaN                           NaN   

      Noroeste PM10  Noreste Ozono  Noreste dioxido de azufre  \
0              37.0           11.0                        4.0   
1              42.0            7.0                        5.0   
2              48.0            7.0                        5.0   
3              52.0            7.0                        5.0   
4              57.0            6.0                        5.0   
...             ...            ...                        ...   
8776            NaN            NaN                        NaN   
8777            NaN            NaN                        NaN   
8778            NaN            NaN                        NaN   
8779            NaN            NaN                        NaN   
8780            NaN            NaN                        NaN   

      Noreste dioxido de nitrogeno  ...  Suroeste PM10  Sureste Ozono  \
0                             20.0  ...           23.0            3.0   
1                             18.0  ...           28.0            7.0   
2                             19.0  ...           33.0            4.0   
3                             18.0  ...           38.0            4.0   
4                             17.0  ...           41.0            5.0   
...                            ...  ...            ...            ...   
8776                           NaN  ...            NaN            NaN   
8777                           NaN  ...            NaN            NaN   
8778                           NaN  ...            NaN            NaN   
8779                           NaN  ...            NaN            NaN   
8780                           NaN  ...            NaN            NaN   

      Sureste dioxido de azufre  Sureste dioxido de nitrogeno  \
0                           3.0                          18.0   
1                           3.0                          16.0   
2                           3.0                          16.0   
3                           3.0                          16.0   
4                           3.0                          15.0   
...                         ...                           ...   
8776                        NaN                           NaN   
8777                        NaN                           NaN   
8778                        NaN                           NaN   
8779                        NaN  

Como se puede observar el archivo del 2013 contiene 8781 filas cuando debería contener 8760 ya que no es bisiesto. Es decir, tiene 21 filas vacías extra. Es necesario eliminar estas filas extra que no aportan información, esto se hará desde DF_all (el que acabamos de limpiar y contiene todos los arvhivos), puesto que son los únicos elementos vacíos que quedan.

In [206]:
#Tamaño antes de limpiar
DF_all.shape

(140277, 27)

In [207]:
#Eliminando filas con fecha vacía
DF_all=DF_all.dropna(axis="rows")

In [208]:
#Tamaño después de limpiar
DF_all.shape

(140256, 27)

In [209]:
#Corroborando que no queden elementos nulos
DF_all.isna().sum()

fecha                            0
hora                             0
noroeste_ozono                   0
noroeste_dioxido_de_azufre       0
noroeste_dioxido_de_nitrogeno    0
noroeste_monoxido_de_carbono     0
noroeste_pm10                    0
noreste_ozono                    0
noreste_dioxido_de_azufre        0
noreste_dioxido_de_nitrogeno     0
noreste_monoxido_de_carbono      0
noreste_pm10                     0
centro_ozono                     0
centro_dioxido_de_azufre         0
centro_dioxido_de_nitrogeno      0
centro_monoxido_de_carbono       0
centro_pm10                      0
suroeste_ozono                   0
suroeste_dioxido_de_azufre       0
suroeste_dioxido_de_nitrogeno    0
suroeste_monoxido_de_carbono     0
suroeste_pm10                    0
sureste_ozono                    0
sureste_dioxido_de_azufre        0
sureste_dioxido_de_nitrogeno     0
sureste_monoxido_de_carbono      0
sureste_pm10                     0
dtype: int64

In [210]:
#Separando la información por contaminante
DF_save = list()
palabra = ['ozono','dioxido_de_azufre','dioxido_de_nitrogeno','monoxido_de_carbono','pm10']
for i in range(0,5):
    DF_save.append(DF_all[['fecha','hora',f'noreste_{palabra[i]}',f'suroeste_{palabra[i]}',f'noroeste_{palabra[i]}',f'sureste_{palabra[i]}',f'centro_{palabra[i]}']])

In [211]:
DF_save[0]

fecha  hora  noreste_ozono  suroeste_ozono  noroeste_ozono  \
0       01/01/2005   1.0           10.0            26.0            19.0   
1       01/01/2005   2.0            7.0            19.0            19.0   
2       01/01/2005   3.0            5.0            23.0            17.0   
3       01/01/2005   4.0            6.0            21.0            16.0   
4       01/01/2005   5.0            8.0            23.0            16.0   
...            ...   ...            ...             ...             ...   
140272  31/12/2020  20.0           14.0            25.0            25.0   
140273  31/12/2020  21.0           11.0            16.0            22.0   
140274  31/12/2020  22.0           10.0             9.0            18.0   
140275  31/12/2020  23.0            9.0            11.0            16.0   
140276  31/12/2020  24.0           11.0            16.0            16.0   

        sureste_ozono  centro_ozono  
0                13.0           4.0  
1                 9.0           4.0  
2                 9.0           4.0  
3                12.0           4.0  
4                19.0           5.0  
...               ...           ...  
140272           21.0          18.0  
140273           21.0          11.0  
140274           22.0           4.0  
140275           29.0           6.0  
140276           20.0           6.0  

[140256 rows x 7 columns]

In [213]:
#Guardamos los datos 

nombres=['O3','SO2','NO2','CO','PM10']
for i in range(0,(len(nombres))):
    DF_save[i].to_csv(f'datos_IMECA/Datos_limpios_imeca/{nombres[i]}.csv',index=False)

# Transformación de Datos

In [214]:
#Lectura de los nuevos archivos creados previamente
CO = pd.read_csv('datos_IMECA/Datos_limpios_imeca/CO.csv', sep = ',')
O3 = pd.read_csv('datos_IMECA/Datos_limpios_imeca/O3.csv', sep = ',')
SO2 = pd.read_csv('datos_IMECA/Datos_limpios_imeca/SO2.csv', sep = ',')
NO2 = pd.read_csv('datos_IMECA/Datos_limpios_imeca/NO2.csv', sep = ',')
PM10 = pd.read_csv('datos_IMECA/Datos_limpios_imeca/PM10.csv', sep = ',')

In [215]:
#Realizar el Casting correspondiente a cada DF
O3['fecha'] = pd.to_datetime(O3['fecha'], unit='ns')
SO2['fecha'] = pd.to_datetime(SO2['fecha'], unit='ns')
NO2['fecha'] = pd.to_datetime(NO2['fecha'], unit='ns')
PM10['fecha'] = pd.to_datetime(PM10['fecha'], unit='ns')

In [216]:
#Comprobación de que se haya realizado correctamente el casting
CO.dtypes
O3.dtypes
SO2.dtypes
NO2.dtypes
PM10.dtypes

fecha            datetime64[ns]
hora                    float64
noreste_pm10            float64
suroeste_pm10           float64
noroeste_pm10           float64
sureste_pm10            float64
centro_pm10             float64
dtype: object

In [144]:
# Aquí traté de agrupar por fechas pero siguen las horas
CO.groupby('fecha')['hora'].value_counts()
O3.groupby('fecha')['hora'].value_counts()
SO2.groupby('fecha')['hora'].value_counts()
NO2.groupby('fecha')['hora'].value_counts()
PM10.groupby('fecha')['hora'].value_counts()

fecha       hora
2005-01-01  1.0     1
            2.0     1
            3.0     1
            4.0     1
            5.0     1
                   ..
2020-12-31  20.0    1
            21.0    1
            22.0    1
            23.0    1
            24.0    1
Name: hora, Length: 140256, dtype: int64

# Agrupacion de datos (PROMEDIO POR MES)

In [217]:
#Elimino hora ya que lo agrupare por meses, en esta agrupacion no se necesita
CO = CO.drop(['hora'], axis=1)
O3 = O3.drop(['hora'], axis=1)
SO2 = SO2.drop(['hora'], axis=1)
NO2 = NO2.drop(['hora'], axis=1)
PM10 = PM10.drop(['hora'], axis=1)

In [218]:
O3.head(5)

fecha  noreste_ozono  suroeste_ozono  noroeste_ozono  sureste_ozono  \
0 2005-01-01           10.0            26.0            19.0           13.0   
1 2005-01-01            7.0            19.0            19.0            9.0   
2 2005-01-01            5.0            23.0            17.0            9.0   
3 2005-01-01            6.0            21.0            16.0           12.0   
4 2005-01-01            8.0            23.0            16.0           19.0   

   centro_ozono  
0           4.0  
1           4.0  
2           4.0  
3           4.0  
4           5.0

In [219]:
CO.fecha = pd.to_datetime(CO.fecha)
O3.fecha = pd.to_datetime(O3.fecha)
SO2.fecha = pd.to_datetime(SO2.fecha)
NO2.fecha = pd.to_datetime(NO2.fecha)
PM10.fecha = pd.to_datetime(PM10.fecha)

In [221]:
O3.head(3)

fecha  noreste_ozono  suroeste_ozono  noroeste_ozono  sureste_ozono  \
0 2005-01-01           10.0            26.0            19.0           13.0   
1 2005-01-01            7.0            19.0            19.0            9.0   
2 2005-01-01            5.0            23.0            17.0            9.0   

   centro_ozono  
0           4.0  
1           4.0  
2           4.0

In [222]:
# TODO: REVISAR ESTA FUNCION
#Se indexa basado en tiempo, lo que permite un mejor manejo con datos temporales   

CO_MESES = CO.resample('M', on='fecha').mean()
O3_MESES = O3.resample('M', on='fecha').mean()
SO2_MESES = SO2.resample('M', on='fecha').mean()
NO2_MESES = NO2.resample('M', on='fecha').mean()
PM10_MESES = PM10.resample('M', on='fecha').mean()

#si quisiera agrupar por año 
O3_YEAR= O3.resample('12M', on='fecha').mean()
#si quisiera agrupar por dia 
O3_DIA= O3.resample('D', on='fecha').mean()

In [223]:
O3_MESES.head()

noreste_ozono  suroeste_ozono  noroeste_ozono  sureste_ozono  \
fecha                                                                      
2005-01-31      34.497312       42.629032       35.766129      40.125000   
2005-02-28      34.537202       42.840774       34.148810      39.645833   
2005-03-31      37.178763       41.713710       34.479839      40.633065   
2005-04-30      43.033333       46.236111       38.872222      43.913889   
2005-05-31      45.501344       50.325269       41.360215      47.463710   

            centro_ozono  
fecha                     
2005-01-31     29.528226  
2005-02-28     28.370536  
2005-03-31     28.346774  
2005-04-30     35.113889  
2005-05-31     38.028226